arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -82,6 +87,14 @@ static const struct z_arm_mpu_partition static_regions[] = {
 		.attr = K_MEM_PARTITION_P_RX_U_RX,
 		},
 #endif /* CONFIG_ARCH_HAS_RAMFUNC_SUPPORT */
+#if defined(CONFIG_CODE_DATA_RELOCATION)
```

Comment:
The reason that this generally works today is that all the "static" regions are located (forced by the linker, actually) in the beginning of SRAM, before the kernel ram (which includes the application memory). This placement is a requirement, currently, for Cortex-M33, because the static memory regions are actually partitioning the background SRAM, because overlapping regions are not allowed. (btw, you cannot currently call configure_static_regions for Cortex-M33 for a region that is not in RAM, unless i am missing something).

In ARMv7-M this is not a requirement, as the MPU static regions do not need to partition the SRAM, and you can configure static regions basically everywhere.

So I am not sure if this solution is generic enough to be accepted. 

@MaureenHelm  @mmahadevan108 did you try to test this PR on a Cortex-M33 with MPU? Especially if the relocation is not on SRAM. If so, I 'd like to hear how this works.

Besides this, I have no more objections for this change.

Reply:
I am running this code on a Cortex-M33. The code is relocated to SRAM.

Reply:
We relocate to ITCM instead of SRAM on i.MX RT1064, which is a Cortex-M7. Not an M33. We talked about it a bit in https://github.com/zephyrproject-rtos/zephyr/pull/31132#discussion_r553349992 when I tried to add an MPU region for ITCM, but we then decided it wasn't necessary.

Reply:
I recall it now, thanks.
In my opinion, we should make sure the solution that is presented here works in a generic way. If not, then we need to restrict it on the ARCHes that support it.

Reply:
To address this issue I have created a new Kconfig called CODE_DATA_RELOCATION_SRAM. The changes made to arm_code_mpu.c file are now using #if defined(CONFIG_CODE_DATA_RELOCATION_SRAM) which allows the MPU entry to be added only when code is relocated to SRAM.